// @typesafe-html5/typescript/src/index.ts
/**
 * Extended HTML Document - TypeScript Classes
 *
 * Auto-generated type-safe TypeScript classes for all HTML5 elements.
 *
 * @package vardumper/extended-htmldocument
 */

// Block Elements
export { Article, type ArticleProps } from './block/article/article';
export { Aside, type AsideProps } from './block/aside/aside';
export { Audio, type AudioProps } from './block/audio/audio';
export { Blockquote, type BlockquoteProps } from './block/blockquote/blockquote';
export { Body, type BodyProps } from './block/body/body';
export { Canvas, type CanvasProps } from './block/canvas/canvas';
export { Caption, type CaptionProps } from './block/caption/caption';
export { Colgroup, type ColgroupProps } from './block/colgroup/colgroup';
export { Datalist, type DatalistProps } from './block/datalist/datalist';
export { Dd, type DdProps } from './block/dd/dd';
export { Del, type DelProps } from './block/del/del';
export { Details, type DetailsProps } from './block/details/details';
export { Dialog, type DialogProps } from './block/dialog/dialog';
export { Div, type DivProps } from './block/div/div';
export { Dl, type DlProps } from './block/dl/dl';
export { Dt, type DtProps } from './block/dt/dt';
export { Embed, type EmbedProps } from './block/embed/embed';
export { Fieldset, type FieldsetProps } from './block/fieldset/fieldset';
export { Figcaption, type FigcaptionProps } from './block/figcaption/figcaption';
export { Figure, type FigureProps } from './block/figure/figure';
export { Footer, type FooterProps } from './block/footer/footer';
export { Form, type FormProps } from './block/form/form';
export { H1, type H1Props } from './block/h1/h1';
export { H2, type H2Props } from './block/h2/h2';
export { H3, type H3Props } from './block/h3/h3';
export { H4, type H4Props } from './block/h4/h4';
export { H5, type H5Props } from './block/h5/h5';
export { H6, type H6Props } from './block/h6/h6';
export { Header, type HeaderProps } from './block/header/header';
export { Hgroup, type HgroupProps } from './block/hgroup/hgroup';
export { Hr, type HrProps } from './block/hr/hr';
export { Html, type HtmlProps } from './block/html/html';
export { Iframe, type IframeProps } from './block/iframe/iframe';
export { Ins, type InsProps } from './block/ins/ins';
export { Legend, type LegendProps } from './block/legend/legend';
export { Li, type LiProps } from './block/li/li';
export { Main, type MainProps } from './block/main/main';
export { Map, type MapProps } from './block/map/map';
export { Menu, type MenuProps } from './block/menu/menu';
export { Nav, type NavProps } from './block/nav/nav';
export { Noscript, type NoscriptProps } from './block/noscript/noscript';
export { Object, type ObjectProps } from './block/object/object';
export { Ol, type OlProps } from './block/ol/ol';
export { Optgroup, type OptgroupProps } from './block/optgroup/optgroup';
export { Option, type OptionProps } from './block/option/option';
export { P, type PProps } from './block/p/p';
export { Picture, type PictureProps } from './block/picture/picture';
export { Pre, type PreProps } from './block/pre/pre';
export { Section, type SectionProps } from './block/section/section';
export { Summary, type SummaryProps } from './block/summary/summary';
export { Table, type TableProps } from './block/table/table';
export { Tbody, type TbodyProps } from './block/tbody/tbody';
export { Td, type TdProps } from './block/td/td';
export { Template, type TemplateProps } from './block/template/template';
export { Tfoot, type TfootProps } from './block/tfoot/tfoot';
export { Th, type ThProps } from './block/th/th';
export { Thead, type TheadProps } from './block/thead/thead';
export { Tr, type TrProps } from './block/tr/tr';
export { Ul, type UlProps } from './block/ul/ul';
export { Video, type VideoProps } from './block/video/video';

// Inline Elements
export { A, type AProps } from './inline/a/a';
export { Abbr, type AbbrProps } from './inline/abbr/abbr';
export { Address, type AddressProps } from './inline/address/address';
export { B, type BProps } from './inline/b/b';
export { Bdi, type BdiProps } from './inline/bdi/bdi';
export { Bdo, type BdoProps } from './inline/bdo/bdo';
export { Button, type ButtonProps } from './inline/button/button';
export { Cite, type CiteProps } from './inline/cite/cite';
export { Code, type CodeProps } from './inline/code/code';
export { Data, type DataProps } from './inline/data/data';
export { Dfn, type DfnProps } from './inline/dfn/dfn';
export { Em, type EmProps } from './inline/em/em';
export { I, type IProps } from './inline/i/i';
export { Img, type ImgProps } from './inline/img/img';
export { Input, type InputProps } from './inline/input/input';
export { Kbd, type KbdProps } from './inline/kbd/kbd';
export { Label, type LabelProps } from './inline/label/label';
export { Mark, type MarkProps } from './inline/mark/mark';
export { Meter, type MeterProps } from './inline/meter/meter';
export { Output, type OutputProps } from './inline/output/output';
export { Progress, type ProgressProps } from './inline/progress/progress';
export { Q, type QProps } from './inline/q/q';
export { Rp, type RpProps } from './inline/rp/rp';
export { Rt, type RtProps } from './inline/rt/rt';
export { Ruby, type RubyProps } from './inline/ruby/ruby';
export { S, type SProps } from './inline/s/s';
export { Samp, type SampProps } from './inline/samp/samp';
export { Select, type SelectProps } from './inline/select/select';
export { Slot, type SlotProps } from './inline/slot/slot';
export { Small, type SmallProps } from './inline/small/small';
export { Span, type SpanProps } from './inline/span/span';
export { Strong, type StrongProps } from './inline/strong/strong';
export { Sub, type SubProps } from './inline/sub/sub';
export { Sup, type SupProps } from './inline/sup/sup';
export { Svg, type SvgProps } from './inline/svg/svg';
export { Textarea, type TextareaProps } from './inline/textarea/textarea';
export { Time, type TimeProps } from './inline/time/time';
export { U, type UProps } from './inline/u/u';
export { Var, type VarProps } from './inline/var/var';

// Void Elements
export { Area, type AreaProps } from './void/area/area';
export { Base, type BaseProps } from './void/base/base';
export { Br, type BrProps } from './void/br/br';
export { Col, type ColProps } from './void/col/col';
export { Head, type HeadProps } from './void/head/head';
export { Link, type LinkProps } from './void/link/link';
export { Meta, type MetaProps } from './void/meta/meta';
export { Param, type ParamProps } from './void/param/param';
export { Script, type ScriptProps } from './void/script/script';
export { Source, type SourceProps } from './void/source/source';
export { Style, type StyleProps } from './void/style/style';
export { Title, type TitleProps } from './void/title/title';
export { Track, type TrackProps } from './void/track/track';
export { Wbr, type WbrProps } from './void/wbr/wbr';
