// @typesafe-html5/typescript/src/void/meta/meta.ts
/**
 * THIS FILE IS AUTOGENERATED. DO NOT EDIT IT.
 *
 * @generated January 2, 2026 15:45:45
 * @component Meta
 * @description 
 */

export interface MetaProps {
  /**
   * Specifies the character encoding for the resource.
   */
  charset?: string | null | undefined;
  /**
   * CSS class names
   */
  className?: string;
  /**
   * Specifies the value associated with the http-equiv or name attribute.
   */
  content?: string | null | undefined;
  /**
   * Mapping of data-* attributes to their values (e.g., {"test":"value"} -> data-test="value")
   */
  'data-attributes'?: Record<string, string> | null | undefined;
  /**
   * Indicates whether the element is hidden
   */
  hidden?: boolean | null | undefined;
  /**
   * Provides an HTTP header for the information/value of the content attribute.
   */
  'http-equiv'?: 'content-language' | 'content-type' | 'default-style' | 'refresh' | null | undefined;
  /**
   * Unique identifier
   */
  id?: string;
  /**
   * Specifies the primary language for the element's content
   */
  lang?: string | null | undefined;
  /**
   * Specifies the name associated with the element. The meaning may vary depending on the context.
   */
  name?: string | null | undefined;
  /**
   * Specifies the content type of the value attribute when the http-equiv attribute is used.
   */
  scheme?: string | null | undefined;
  /**
   * Represents a title or tooltip for the element
   */
  title?: string | null | undefined;
}

/**
 * Meta - 
 */
export class Meta {
  private element: HTMLElement;

  constructor(props: MetaProps = {}) {
    this.element = document.createElement('meta');
    this.applyProps(props);
  }

  private applyProps(props: MetaProps): void {
    if (props.charset !== undefined) {
      this.element.setAttribute('charset', String(props.charset));
    }
    if (props.className !== undefined) {
      this.element.setAttribute('className', String(props.className));
    }
    if (props.content !== undefined) {
      this.element.setAttribute('content', String(props.content));
    }
    if (props['data-attributes'] !== undefined) {
      const __map = props['data-attributes'];
      if (__map !== null && __map !== undefined) {
        for (const __k in __map) {
          this.element.setAttribute(`data-${__k}`, String((__map as any)[__k]));
        }
      }
    }
    if (props.hidden !== undefined) {
      this.element.setAttribute('hidden', String(props.hidden));
    }
    if (props['http-equiv'] !== undefined) {
      this.setHttpEquiv(props['http-equiv']);
    }
    if (props.id !== undefined) {
      this.element.setAttribute('id', String(props.id));
    }
    if (props.lang !== undefined) {
      this.element.setAttribute('lang', String(props.lang));
    }
    if (props.name !== undefined) {
      this.element.setAttribute('name', String(props.name));
    }
    if (props.scheme !== undefined) {
      this.element.setAttribute('scheme', String(props.scheme));
    }
    if (props.title !== undefined) {
      this.element.setAttribute('title', String(props.title));
    }
  }

  setCharset(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('charset');
    } else {
      this.element.setAttribute('charset', String(value));
    }
    return this;
  }

  setClassName(value: string): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('className');
    } else {
      this.element.setAttribute('className', String(value));
    }
    return this;
  }

  setContent(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('content');
    } else {
      this.element.setAttribute('content', String(value));
    }
    return this;
  }

  setDataAttributes(value: Record<string, string>): this {
    if (value === null || value === undefined) {
      // Remove all data-* attributes (best-effort)
      for (const __a of Array.from(this.element.attributes)) {
        if ((__a as Attr).name.startsWith('data-')) { (this.element as HTMLElement).removeAttribute((__a as Attr).name); }
      }
    } else {
      for (const __k in value as Record<string, string>) {
        this.element.setAttribute(`data-${__k}`, String((value as any)[__k]));
      }
    }
    return this;
  }

  setHidden(value: boolean | null | undefined): this {
    if (value === true) {
      this.element.setAttribute('hidden', '');
    } else {
      this.element.removeAttribute('hidden');
    }
    return this;
  }

  setHttpEquiv(value: 'content-language' | 'content-type' | 'default-style' | 'refresh' | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('http-equiv');
    } else {
      this.element.setAttribute('http-equiv', String(value));
    }
    return this;
  }

  setId(value: string): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('id');
    } else {
      this.element.setAttribute('id', String(value));
    }
    return this;
  }

  setLang(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('lang');
    } else {
      this.element.setAttribute('lang', String(value));
    }
    return this;
  }

  setName(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('name');
    } else {
      this.element.setAttribute('name', String(value));
    }
    return this;
  }

  setScheme(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('scheme');
    } else {
      this.element.setAttribute('scheme', String(value));
    }
    return this;
  }

  setTitle(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('title');
    } else {
      this.element.setAttribute('title', String(value));
    }
    return this;
  }

  getElement(): HTMLElement {
    return this.element;
  }
}
