// @typesafe-html5/typescript/src/inline/progress/progress.ts
/**
 * THIS FILE IS AUTOGENERATED. DO NOT EDIT IT.
 *
 * @generated January 2, 2026 15:45:45
 * @component Progress
 * @description 
 */

export interface ProgressProps {
  /**
   * Child content or elements
   */
  children?: string | HTMLElement | (string | HTMLElement)[];
  /**
   * Keyboard shortcut to activate or focus the element
   */
  accessKey?: string;
  /**
   * specifies a shortcut key (or keys) to activate or focus an element
   */
  accesskey?: string | null | undefined;
  /**
   * Mapping of Alpine.js directives (shorthand or full) to their values (e.g., {@click: "do()"} or {"x-on:click":"do()"})
   */
  'alpine-attributes'?: Record<string, string> | null | undefined;
  /**
   * Indicates whether assistive technologies should present the entire region as a whole when changes occur.
   */
  'aria-atomic'?: 'false' | 'true' | boolean | null | undefined;
  /**
   * References an element that provides additional details about the current element.
   */
  'aria-details'?: string | null | undefined;
  /**
   * Defines keyboard shortcuts available for the element.
   */
  'aria-keyshortcuts'?: string | null | undefined;
  /**
   * Defines how updates to the element should be announced to screen readers.
   */
  'aria-live'?: 'off' | 'polite' | 'assertive' | null | undefined;
  /**
   * Indicates what content changes should be announced in a live region.
   */
  'aria-relevant'?: 'additions' | 'removals' | 'text' | 'all' | 'additions text' | null | undefined;
  /**
   * Provides a human-readable custom role description for assistive technologies.
   */
  'aria-roledescription'?: string | null | undefined;
  /**
   * Defines the maximum value for a range input.
   */
  'aria-valuemax'?: number | null | undefined;
  /**
   * Defines the minimum value for a range input.
   */
  'aria-valuemin'?: number | null | undefined;
  /**
   * Specifies the current value for a range input.
   */
  'aria-valuenow'?: number | null | undefined;
  /**
   * Provides a human-readable representation of the current value.
   */
  'aria-valuetext'?: string | null | undefined;
  /**
   * Represents the autocapitalize behavior of the element
   */
  autocapitalize?: 'none' | 'sentences' | 'words' | 'characters' | null | undefined;
  /**
   * Indicates whether the element is hidden
   */
  autofocus?: boolean | null | undefined;
  /**
   * CSS class names
   */
  className?: string;
  /**
   * Whether the element is editable
   */
  contentEditable?: boolean | "true" | "false" | "inherit";
  /**
   * Indicates whether the element can be edited in place
   */
  contenteditable?: 'true' | 'false' | 'inherit' | boolean | null | undefined;
  /**
   * Mapping of data-* attributes to their values (e.g., {"test":"value"} -> data-test="value")
   */
  'data-attributes'?: Record<string, string> | null | undefined;
  /**
   * Represents the text direction of the element
   */
  dir?: 'ltr' | 'rtl' | 'auto' | null | undefined;
  /**
   * Indicates whether the element is draggable
   */
  draggable?: boolean | null | undefined;
  /**
   * Indicates whether the element is hidden
   */
  hidden?: boolean | null | undefined;
  /**
   * Unique identifier
   */
  id?: string;
  /**
   * used to specify the data entry mode for an input. It helps guide on-screen keyboards (especially on mobile devices) to show the appropriate layout for the expected input type
   */
  inputmode?: 'none' | 'text' | 'decimal' | 'numeric' | 'email' | 'tel' | 'url' | 'search' | null | undefined;
  /**
   * Specifies the primary language for the element's content
   */
  lang?: string | null | undefined;
  /**
   * Specifies the maximum value for an input element, meter, or progress element.
   */
  max?: number | null | undefined;
  /**
   * Represents a slot in a shadow DOM
   */
  slot?: string | null | undefined;
  /**
   * Represents the spellchecking behavior of the element
   */
  spellcheck?: 'true' | 'false' | boolean | null | undefined;
  /**
   * Represents the CSS inline style of the element
   */
  style?: string | null | undefined;
  /**
   * Tab order position
   */
  tabIndex?: number;
  /**
   * Represents a tab order of the element
   */
  tabindex?: number | null | undefined;
  /**
   * Represents a title or tooltip for the element
   */
  title?: string | null | undefined;
  /**
   * used to tell user agents whether the content should be translated.
   */
  translate?: 'yes' | 'no' | null | undefined;
  /**
   * Specifies the value associated with the element. The meaning and usage may vary depending on the element type.
   */
  value?: string | null | undefined;
}

/**
 * Progress - 
 */
export class Progress {
  private element: HTMLElement;

  constructor(props: ProgressProps = {}) {
    this.element = document.createElement('progress');
    this.applyProps(props);
  }

  private applyProps(props: ProgressProps): void {
    if (props.children !== undefined) {
      this.setChildren(props.children);
    }
    if (props.accessKey !== undefined) {
      this.element.setAttribute('accessKey', String(props.accessKey));
    }
    if (props.accesskey !== undefined) {
      this.element.setAttribute('accesskey', String(props.accesskey));
    }
    if (props['alpine-attributes'] !== undefined) {
      const __map = props['alpine-attributes'];
      if (__map !== null && __map !== undefined) {
        for (const __k in __map) {
          const __v = (__map as any)[__k];
          let __attr = __k;
          if (__k.startsWith('@')) { __attr = 'x-on:' + __k.slice(1); }
          else if (__k.startsWith(':')) { __attr = 'x-bind:' + __k.slice(1); }
          else if (__k.startsWith('.')) { __attr = 'x-model' + __k; }
          else if (!__k.startsWith('x-') && ['show','text','html','if','for','cloak','init','data','effect','ignore','ref','transition','teleport'].includes(__k)) { __attr = 'x-' + __k; }
          this.element.setAttribute(__attr, String(__v));
        }
      }
    }
    if (props['aria-atomic'] !== undefined) {
      this.setAriaAtomic(props['aria-atomic']);
    }
    if (props['aria-details'] !== undefined) {
      this.element.setAttribute('aria-details', String(props['aria-details']));
    }
    if (props['aria-keyshortcuts'] !== undefined) {
      this.element.setAttribute('aria-keyshortcuts', String(props['aria-keyshortcuts']));
    }
    if (props['aria-live'] !== undefined) {
      this.setAriaLive(props['aria-live']);
    }
    if (props['aria-relevant'] !== undefined) {
      this.setAriaRelevant(props['aria-relevant']);
    }
    if (props['aria-roledescription'] !== undefined) {
      this.element.setAttribute('aria-roledescription', String(props['aria-roledescription']));
    }
    if (props['aria-valuemax'] !== undefined) {
      this.element.setAttribute('aria-valuemax', String(props['aria-valuemax']));
    }
    if (props['aria-valuemin'] !== undefined) {
      this.element.setAttribute('aria-valuemin', String(props['aria-valuemin']));
    }
    if (props['aria-valuenow'] !== undefined) {
      this.element.setAttribute('aria-valuenow', String(props['aria-valuenow']));
    }
    if (props['aria-valuetext'] !== undefined) {
      this.element.setAttribute('aria-valuetext', String(props['aria-valuetext']));
    }
    if (props.autocapitalize !== undefined) {
      this.setAutocapitalize(props.autocapitalize);
    }
    if (props.autofocus !== undefined) {
      this.element.setAttribute('autofocus', String(props.autofocus));
    }
    if (props.className !== undefined) {
      this.element.setAttribute('className', String(props.className));
    }
    if (props.contentEditable !== undefined) {
      this.element.setAttribute('contentEditable', String(props.contentEditable));
    }
    if (props.contenteditable !== undefined) {
      this.setContenteditable(props.contenteditable);
    }
    if (props['data-attributes'] !== undefined) {
      const __map = props['data-attributes'];
      if (__map !== null && __map !== undefined) {
        for (const __k in __map) {
          this.element.setAttribute(`data-${__k}`, String((__map as any)[__k]));
        }
      }
    }
    if (props.dir !== undefined) {
      this.setDir(props.dir);
    }
    if (props.draggable !== undefined) {
      this.element.setAttribute('draggable', String(props.draggable));
    }
    if (props.hidden !== undefined) {
      this.element.setAttribute('hidden', String(props.hidden));
    }
    if (props.id !== undefined) {
      this.element.setAttribute('id', String(props.id));
    }
    if (props.inputmode !== undefined) {
      this.setInputmode(props.inputmode);
    }
    if (props.lang !== undefined) {
      this.element.setAttribute('lang', String(props.lang));
    }
    if (props.max !== undefined) {
      this.element.setAttribute('max', String(props.max));
    }
    if (props.slot !== undefined) {
      this.element.setAttribute('slot', String(props.slot));
    }
    if (props.spellcheck !== undefined) {
      this.setSpellcheck(props.spellcheck);
    }
    if (props.style !== undefined) {
      this.element.setAttribute('style', String(props.style));
    }
    if (props.tabIndex !== undefined) {
      this.element.setAttribute('tabIndex', String(props.tabIndex));
    }
    if (props.tabindex !== undefined) {
      this.element.setAttribute('tabindex', String(props.tabindex));
    }
    if (props.title !== undefined) {
      this.element.setAttribute('title', String(props.title));
    }
    if (props.translate !== undefined) {
      this.setTranslate(props.translate);
    }
    if (props.value !== undefined) {
      this.element.setAttribute('value', String(props.value));
    }
  }

  setAccessKey(value: string): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('accessKey');
    } else {
      this.element.setAttribute('accessKey', String(value));
    }
    return this;
  }

  setAccesskey(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('accesskey');
    } else {
      this.element.setAttribute('accesskey', String(value));
    }
    return this;
  }

  setAlpineAttributes(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      // Remove all x-* Alpine attributes (best-effort)
      for (const __a of Array.from(this.element.attributes)) {
        if ((__a as Attr).name.startsWith('x-')) { (this.element as HTMLElement).removeAttribute((__a as Attr).name); }
      }
    } else {
      for (const __k in value as Record<string, string>) {
        let __attr = __k;
        if (__k.startsWith('@')) { __attr = 'x-on:' + __k.slice(1); }
        else if (__k.startsWith(':')) { __attr = 'x-bind:' + __k.slice(1); }
        else if (__k.startsWith('.')) { __attr = 'x-model' + __k; }
        else if (!__k.startsWith('x-') && ['show','text','html','if','for','cloak','init','data','effect','ignore','ref','transition','teleport'].includes(__k)) { __attr = 'x-' + __k; }
        this.element.setAttribute(__attr, String((value as any)[__k]));
      }
    }
    return this;
  }

  setAriaAtomic(value: 'false' | 'true' | boolean | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('aria-atomic');
    } else {
      this.element.setAttribute('aria-atomic', String(value));
    }
    return this;
  }

  setAriaDetails(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('aria-details');
    } else {
      this.element.setAttribute('aria-details', String(value));
    }
    return this;
  }

  setAriaKeyshortcuts(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('aria-keyshortcuts');
    } else {
      this.element.setAttribute('aria-keyshortcuts', String(value));
    }
    return this;
  }

  setAriaLive(value: 'off' | 'polite' | 'assertive' | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('aria-live');
    } else {
      this.element.setAttribute('aria-live', String(value));
    }
    return this;
  }

  setAriaRelevant(value: 'additions' | 'removals' | 'text' | 'all' | 'additions text' | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('aria-relevant');
    } else {
      this.element.setAttribute('aria-relevant', String(value));
    }
    return this;
  }

  setAriaRoledescription(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('aria-roledescription');
    } else {
      this.element.setAttribute('aria-roledescription', String(value));
    }
    return this;
  }

  setAriaValuemax(value: number | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('aria-valuemax');
    } else {
      this.element.setAttribute('aria-valuemax', String(value));
    }
    return this;
  }

  setAriaValuemin(value: number | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('aria-valuemin');
    } else {
      this.element.setAttribute('aria-valuemin', String(value));
    }
    return this;
  }

  setAriaValuenow(value: number | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('aria-valuenow');
    } else {
      this.element.setAttribute('aria-valuenow', String(value));
    }
    return this;
  }

  setAriaValuetext(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('aria-valuetext');
    } else {
      this.element.setAttribute('aria-valuetext', String(value));
    }
    return this;
  }

  setAutocapitalize(value: 'none' | 'sentences' | 'words' | 'characters' | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('autocapitalize');
    } else {
      this.element.setAttribute('autocapitalize', String(value));
    }
    return this;
  }

  setAutofocus(value: boolean | null | undefined): this {
    if (value === true) {
      this.element.setAttribute('autofocus', '');
    } else {
      this.element.removeAttribute('autofocus');
    }
    return this;
  }

  setClassName(value: string): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('className');
    } else {
      this.element.setAttribute('className', String(value));
    }
    return this;
  }

  setContentEditable(value: boolean | "true" | "false" | "inherit"): this {
    if (value === true) {
      this.element.setAttribute('contentEditable', '');
    } else {
      this.element.removeAttribute('contentEditable');
    }
    return this;
  }

  setContenteditable(value: 'true' | 'false' | 'inherit' | boolean | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('contenteditable');
    } else {
      this.element.setAttribute('contenteditable', String(value));
    }
    return this;
  }

  setDataAttributes(value: Record<string, string>): this {
    if (value === null || value === undefined) {
      // Remove all data-* attributes (best-effort)
      for (const __a of Array.from(this.element.attributes)) {
        if ((__a as Attr).name.startsWith('data-')) { (this.element as HTMLElement).removeAttribute((__a as Attr).name); }
      }
    } else {
      for (const __k in value as Record<string, string>) {
        this.element.setAttribute(`data-${__k}`, String((value as any)[__k]));
      }
    }
    return this;
  }

  setDir(value: 'ltr' | 'rtl' | 'auto' | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('dir');
    } else {
      this.element.setAttribute('dir', String(value));
    }
    return this;
  }

  setDraggable(value: boolean | null | undefined): this {
    if (value === true) {
      this.element.setAttribute('draggable', '');
    } else {
      this.element.removeAttribute('draggable');
    }
    return this;
  }

  setHidden(value: boolean | null | undefined): this {
    if (value === true) {
      this.element.setAttribute('hidden', '');
    } else {
      this.element.removeAttribute('hidden');
    }
    return this;
  }

  setId(value: string): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('id');
    } else {
      this.element.setAttribute('id', String(value));
    }
    return this;
  }

  setInputmode(value: 'none' | 'text' | 'decimal' | 'numeric' | 'email' | 'tel' | 'url' | 'search' | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('inputmode');
    } else {
      this.element.setAttribute('inputmode', String(value));
    }
    return this;
  }

  setLang(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('lang');
    } else {
      this.element.setAttribute('lang', String(value));
    }
    return this;
  }

  setMax(value: number | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('max');
    } else {
      this.element.setAttribute('max', String(value));
    }
    return this;
  }

  setSlot(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('slot');
    } else {
      this.element.setAttribute('slot', String(value));
    }
    return this;
  }

  setSpellcheck(value: 'true' | 'false' | boolean | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('spellcheck');
    } else {
      this.element.setAttribute('spellcheck', String(value));
    }
    return this;
  }

  setStyle(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('style');
    } else {
      this.element.setAttribute('style', String(value));
    }
    return this;
  }

  setTabIndex(value: number): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('tabIndex');
    } else {
      this.element.setAttribute('tabIndex', String(value));
    }
    return this;
  }

  setTabindex(value: number | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('tabindex');
    } else {
      this.element.setAttribute('tabindex', String(value));
    }
    return this;
  }

  setTitle(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('title');
    } else {
      this.element.setAttribute('title', String(value));
    }
    return this;
  }

  setTranslate(value: 'yes' | 'no' | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('translate');
    } else {
      this.element.setAttribute('translate', String(value));
    }
    return this;
  }

  setValue(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('value');
    } else {
      this.element.setAttribute('value', String(value));
    }
    return this;
  }

  setChildren(children: string | HTMLElement | (string | HTMLElement)[]): this {
    // Clear existing children
    while (this.element.firstChild) {
      this.element.removeChild(this.element.firstChild);
    }
    
    if (typeof children === 'string') {
      this.element.textContent = children;
    } else if (Array.isArray(children)) {
      children.forEach(child => {
        if (typeof child === 'string') {
          this.element.appendChild(document.createTextNode(child));
        } else {
          this.element.appendChild(child);
        }
      });
    } else {
      this.element.appendChild(children);
    }
    return this;
  }

  getElement(): HTMLElement {
    return this.element;
  }
}
