// @typesafe-html5/typescript/src/void/source/source.ts
/**
 * THIS FILE IS AUTOGENERATED. DO NOT EDIT IT.
 *
 * @generated January 2, 2026 15:45:45
 * @component Source
 * @description 
 */

export interface SourceProps {
  /**
   * Mapping of Alpine.js directives (shorthand or full) to their values (e.g., {@click: "do()"} or {"x-on:click":"do()"})
   */
  'alpine-attributes'?: Record<string, string> | null | undefined;
  /**
   * CSS class names
   */
  className?: string;
  /**
   * Mapping of data-* attributes to their values (e.g., {"test":"value"} -> data-test="value")
   */
  'data-attributes'?: Record<string, string> | null | undefined;
  /**
   * Indicates whether the element is hidden
   */
  hidden?: boolean | null | undefined;
  /**
   * Unique identifier
   */
  id?: string;
  /**
   * Specifies the media type for which the linked resource or style sheet is intended.
   */
  media?: string | null | undefined;
  /**
   * Specifies the sizes of the images or icons for different display/window sizes.
   */
  sizes?: string | null | undefined;
  /**
   * Specifies the URL of the external resource to be embedded or referenced.
   */
  src?: string | null | undefined;
  /**
   * Specifies a list of possible image sources for the browser to use.
   */
  srcset?: string | null | undefined;
  /**
   * Specifies the media type of the linked resource.
   */
  type?: string | null | undefined;
}

/**
 * Source - 
 */
export class Source {
  private element: HTMLElement;

  constructor(props: SourceProps = {}) {
    this.element = document.createElement('source');
    this.applyProps(props);
  }

  private applyProps(props: SourceProps): void {
    if (props['alpine-attributes'] !== undefined) {
      const __map = props['alpine-attributes'];
      if (__map !== null && __map !== undefined) {
        for (const __k in __map) {
          const __v = (__map as any)[__k];
          let __attr = __k;
          if (__k.startsWith('@')) { __attr = 'x-on:' + __k.slice(1); }
          else if (__k.startsWith(':')) { __attr = 'x-bind:' + __k.slice(1); }
          else if (__k.startsWith('.')) { __attr = 'x-model' + __k; }
          else if (!__k.startsWith('x-') && ['show','text','html','if','for','cloak','init','data','effect','ignore','ref','transition','teleport'].includes(__k)) { __attr = 'x-' + __k; }
          this.element.setAttribute(__attr, String(__v));
        }
      }
    }
    if (props.className !== undefined) {
      this.element.setAttribute('className', String(props.className));
    }
    if (props['data-attributes'] !== undefined) {
      const __map = props['data-attributes'];
      if (__map !== null && __map !== undefined) {
        for (const __k in __map) {
          this.element.setAttribute(`data-${__k}`, String((__map as any)[__k]));
        }
      }
    }
    if (props.hidden !== undefined) {
      this.element.setAttribute('hidden', String(props.hidden));
    }
    if (props.id !== undefined) {
      this.element.setAttribute('id', String(props.id));
    }
    if (props.media !== undefined) {
      this.element.setAttribute('media', String(props.media));
    }
    if (props.sizes !== undefined) {
      this.element.setAttribute('sizes', String(props.sizes));
    }
    if (props.src !== undefined) {
      this.element.setAttribute('src', String(props.src));
    }
    if (props.srcset !== undefined) {
      this.element.setAttribute('srcset', String(props.srcset));
    }
    if (props.type !== undefined) {
      this.element.setAttribute('type', String(props.type));
    }
  }

  setAlpineAttributes(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      // Remove all x-* Alpine attributes (best-effort)
      for (const __a of Array.from(this.element.attributes)) {
        if ((__a as Attr).name.startsWith('x-')) { (this.element as HTMLElement).removeAttribute((__a as Attr).name); }
      }
    } else {
      for (const __k in value as Record<string, string>) {
        let __attr = __k;
        if (__k.startsWith('@')) { __attr = 'x-on:' + __k.slice(1); }
        else if (__k.startsWith(':')) { __attr = 'x-bind:' + __k.slice(1); }
        else if (__k.startsWith('.')) { __attr = 'x-model' + __k; }
        else if (!__k.startsWith('x-') && ['show','text','html','if','for','cloak','init','data','effect','ignore','ref','transition','teleport'].includes(__k)) { __attr = 'x-' + __k; }
        this.element.setAttribute(__attr, String((value as any)[__k]));
      }
    }
    return this;
  }

  setClassName(value: string): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('className');
    } else {
      this.element.setAttribute('className', String(value));
    }
    return this;
  }

  setDataAttributes(value: Record<string, string>): this {
    if (value === null || value === undefined) {
      // Remove all data-* attributes (best-effort)
      for (const __a of Array.from(this.element.attributes)) {
        if ((__a as Attr).name.startsWith('data-')) { (this.element as HTMLElement).removeAttribute((__a as Attr).name); }
      }
    } else {
      for (const __k in value as Record<string, string>) {
        this.element.setAttribute(`data-${__k}`, String((value as any)[__k]));
      }
    }
    return this;
  }

  setHidden(value: boolean | null | undefined): this {
    if (value === true) {
      this.element.setAttribute('hidden', '');
    } else {
      this.element.removeAttribute('hidden');
    }
    return this;
  }

  setId(value: string): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('id');
    } else {
      this.element.setAttribute('id', String(value));
    }
    return this;
  }

  setMedia(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('media');
    } else {
      this.element.setAttribute('media', String(value));
    }
    return this;
  }

  setSizes(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('sizes');
    } else {
      this.element.setAttribute('sizes', String(value));
    }
    return this;
  }

  setSrc(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('src');
    } else {
      this.element.setAttribute('src', String(value));
    }
    return this;
  }

  setSrcset(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('srcset');
    } else {
      this.element.setAttribute('srcset', String(value));
    }
    return this;
  }

  setType(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('type');
    } else {
      this.element.setAttribute('type', String(value));
    }
    return this;
  }

  getElement(): HTMLElement {
    return this.element;
  }
}
