// @typesafe-html5/typescript/src/void/br/br.ts
/**
 * THIS FILE IS AUTOGENERATED. DO NOT EDIT IT.
 *
 * @generated January 2, 2026 15:45:45
 * @component Br
 * @description 
 */

export interface BrProps {
  /**
   * Mapping of Alpine.js directives (shorthand or full) to their values (e.g., {@click: "do()"} or {"x-on:click":"do()"})
   */
  'alpine-attributes'?: Record<string, string> | null | undefined;
  /**
   * CSS class names
   */
  className?: string;
  /**
   * Mapping of data-* attributes to their values (e.g., {"test":"value"} -> data-test="value")
   */
  'data-attributes'?: Record<string, string> | null | undefined;
  /**
   * Represents the text direction of the element
   */
  dir?: 'ltr' | 'rtl' | 'auto' | null | undefined;
  /**
   * Indicates whether the element is hidden
   */
  hidden?: boolean | null | undefined;
  /**
   * Unique identifier
   */
  id?: string;
  /**
   * Represents the CSS inline style of the element
   */
  style?: string | null | undefined;
}

/**
 * Br - 
 */
export class Br {
  private element: HTMLElement;

  constructor(props: BrProps = {}) {
    this.element = document.createElement('br');
    this.applyProps(props);
  }

  private applyProps(props: BrProps): void {
    if (props['alpine-attributes'] !== undefined) {
      const __map = props['alpine-attributes'];
      if (__map !== null && __map !== undefined) {
        for (const __k in __map) {
          const __v = (__map as any)[__k];
          let __attr = __k;
          if (__k.startsWith('@')) { __attr = 'x-on:' + __k.slice(1); }
          else if (__k.startsWith(':')) { __attr = 'x-bind:' + __k.slice(1); }
          else if (__k.startsWith('.')) { __attr = 'x-model' + __k; }
          else if (!__k.startsWith('x-') && ['show','text','html','if','for','cloak','init','data','effect','ignore','ref','transition','teleport'].includes(__k)) { __attr = 'x-' + __k; }
          this.element.setAttribute(__attr, String(__v));
        }
      }
    }
    if (props.className !== undefined) {
      this.element.setAttribute('className', String(props.className));
    }
    if (props['data-attributes'] !== undefined) {
      const __map = props['data-attributes'];
      if (__map !== null && __map !== undefined) {
        for (const __k in __map) {
          this.element.setAttribute(`data-${__k}`, String((__map as any)[__k]));
        }
      }
    }
    if (props.dir !== undefined) {
      this.setDir(props.dir);
    }
    if (props.hidden !== undefined) {
      this.element.setAttribute('hidden', String(props.hidden));
    }
    if (props.id !== undefined) {
      this.element.setAttribute('id', String(props.id));
    }
    if (props.style !== undefined) {
      this.element.setAttribute('style', String(props.style));
    }
  }

  setAlpineAttributes(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      // Remove all x-* Alpine attributes (best-effort)
      for (const __a of Array.from(this.element.attributes)) {
        if ((__a as Attr).name.startsWith('x-')) { (this.element as HTMLElement).removeAttribute((__a as Attr).name); }
      }
    } else {
      for (const __k in value as Record<string, string>) {
        let __attr = __k;
        if (__k.startsWith('@')) { __attr = 'x-on:' + __k.slice(1); }
        else if (__k.startsWith(':')) { __attr = 'x-bind:' + __k.slice(1); }
        else if (__k.startsWith('.')) { __attr = 'x-model' + __k; }
        else if (!__k.startsWith('x-') && ['show','text','html','if','for','cloak','init','data','effect','ignore','ref','transition','teleport'].includes(__k)) { __attr = 'x-' + __k; }
        this.element.setAttribute(__attr, String((value as any)[__k]));
      }
    }
    return this;
  }

  setClassName(value: string): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('className');
    } else {
      this.element.setAttribute('className', String(value));
    }
    return this;
  }

  setDataAttributes(value: Record<string, string>): this {
    if (value === null || value === undefined) {
      // Remove all data-* attributes (best-effort)
      for (const __a of Array.from(this.element.attributes)) {
        if ((__a as Attr).name.startsWith('data-')) { (this.element as HTMLElement).removeAttribute((__a as Attr).name); }
      }
    } else {
      for (const __k in value as Record<string, string>) {
        this.element.setAttribute(`data-${__k}`, String((value as any)[__k]));
      }
    }
    return this;
  }

  setDir(value: 'ltr' | 'rtl' | 'auto' | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('dir');
    } else {
      this.element.setAttribute('dir', String(value));
    }
    return this;
  }

  setHidden(value: boolean | null | undefined): this {
    if (value === true) {
      this.element.setAttribute('hidden', '');
    } else {
      this.element.removeAttribute('hidden');
    }
    return this;
  }

  setId(value: string): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('id');
    } else {
      this.element.setAttribute('id', String(value));
    }
    return this;
  }

  setStyle(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('style');
    } else {
      this.element.setAttribute('style', String(value));
    }
    return this;
  }

  getElement(): HTMLElement {
    return this.element;
  }
}
