// @typesafe-html5/typescript/src/void/style/style.ts
/**
 * THIS FILE IS AUTOGENERATED. DO NOT EDIT IT.
 *
 * @generated January 2, 2026 15:45:45
 * @component Style
 * @description 
 */

export interface StyleProps {
  /**
   * Child content or elements
   */
  children?: string | HTMLElement | (string | HTMLElement)[];
  /**
   * CSS class names
   */
  className?: string;
  /**
   * Mapping of data-* attributes to their values (e.g., {"test":"value"} -> data-test="value")
   */
  'data-attributes'?: Record<string, string> | null | undefined;
  /**
   * Represents the text direction of the element
   */
  dir?: 'ltr' | 'rtl' | 'auto' | null | undefined;
  /**
   * Unique identifier
   */
  id?: string;
  /**
   * Specifies the primary language for the element's content
   */
  lang?: string | null | undefined;
  /**
   * Specifies the media type for which the linked resource or style sheet is intended.
   */
  media?: string | null | undefined;
  /**
   * Specifies a cryptographic nonce that can be used in Content Security Policy (CSP) checks.
   */
  nonce?: string | null | undefined;
  /**
   * Represents a title or tooltip for the element
   */
  title?: string | null | undefined;
  /**
   * Specifies the media type of the inline styles.
   */
  type?: 'text/css' | null | undefined;
}

/**
 * Style - 
 */
export class Style {
  private element: HTMLElement;

  constructor(props: StyleProps = {}) {
    this.element = document.createElement('style');
    this.applyProps(props);
  }

  private applyProps(props: StyleProps): void {
    if (props.children !== undefined) {
      this.setChildren(props.children);
    }
    if (props.className !== undefined) {
      this.element.setAttribute('className', String(props.className));
    }
    if (props['data-attributes'] !== undefined) {
      const __map = props['data-attributes'];
      if (__map !== null && __map !== undefined) {
        for (const __k in __map) {
          this.element.setAttribute(`data-${__k}`, String((__map as any)[__k]));
        }
      }
    }
    if (props.dir !== undefined) {
      this.setDir(props.dir);
    }
    if (props.id !== undefined) {
      this.element.setAttribute('id', String(props.id));
    }
    if (props.lang !== undefined) {
      this.element.setAttribute('lang', String(props.lang));
    }
    if (props.media !== undefined) {
      this.element.setAttribute('media', String(props.media));
    }
    if (props.nonce !== undefined) {
      this.element.setAttribute('nonce', String(props.nonce));
    }
    if (props.title !== undefined) {
      this.element.setAttribute('title', String(props.title));
    }
    if (props.type !== undefined) {
      this.setType(props.type);
    }
  }

  setClassName(value: string): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('className');
    } else {
      this.element.setAttribute('className', String(value));
    }
    return this;
  }

  setDataAttributes(value: Record<string, string>): this {
    if (value === null || value === undefined) {
      // Remove all data-* attributes (best-effort)
      for (const __a of Array.from(this.element.attributes)) {
        if ((__a as Attr).name.startsWith('data-')) { (this.element as HTMLElement).removeAttribute((__a as Attr).name); }
      }
    } else {
      for (const __k in value as Record<string, string>) {
        this.element.setAttribute(`data-${__k}`, String((value as any)[__k]));
      }
    }
    return this;
  }

  setDir(value: 'ltr' | 'rtl' | 'auto' | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('dir');
    } else {
      this.element.setAttribute('dir', String(value));
    }
    return this;
  }

  setId(value: string): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('id');
    } else {
      this.element.setAttribute('id', String(value));
    }
    return this;
  }

  setLang(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('lang');
    } else {
      this.element.setAttribute('lang', String(value));
    }
    return this;
  }

  setMedia(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('media');
    } else {
      this.element.setAttribute('media', String(value));
    }
    return this;
  }

  setNonce(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('nonce');
    } else {
      this.element.setAttribute('nonce', String(value));
    }
    return this;
  }

  setTitle(value: string | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('title');
    } else {
      this.element.setAttribute('title', String(value));
    }
    return this;
  }

  setType(value: 'text/css' | null | undefined): this {
    if (value === null || value === undefined) {
      this.element.removeAttribute('type');
    } else {
      this.element.setAttribute('type', String(value));
    }
    return this;
  }

  setChildren(children: string | HTMLElement | (string | HTMLElement)[]): this {
    // Clear existing children
    while (this.element.firstChild) {
      this.element.removeChild(this.element.firstChild);
    }
    
    if (typeof children === 'string') {
      this.element.textContent = children;
    } else if (Array.isArray(children)) {
      children.forEach(child => {
        if (typeof child === 'string') {
          this.element.appendChild(document.createTextNode(child));
        } else {
          this.element.appendChild(child);
        }
      });
    } else {
      this.element.appendChild(children);
    }
    return this;
  }

  getElement(): HTMLElement {
    return this.element;
  }
}
